CIRC-593: Backend: Display enumeration, chronology, volume in loans contexts.

VRohach opened the pull request:
<!--
  If you have a relevant JIRA issue number, please put it in the issue title.
  Example: MODORDERS-70 Orders schema updates

  TL;DR
    - https://www.youtube.com/watch?v=5aHmO_S8FQ4
    - http://www.olitreadwell.com/2016/05/22/how-to-write-great-pull-requests/
    - https://www.atlassian.com/blog/git/written-unwritten-guide-pull-requests
-->

JIRA-ticket: https://issues.folio.org/browse/CIRC-593

## Purpose
<!--
  Why are you making this change? There is nothing more important
  to provide to the reviewer and to future readers than the cause
  that gave rise to this pull request. Be careful to avoid circular
  statements like "the purpose is to update the schema." and
  instead provide an explanation like "there is more data to be provided and stored for Purchase Orders 
  which is currently missing in the schema"

  The purpose may seem self-evident to you now, but the standard to
  hold yourself to should be "can a developer parachuting into this
  project reconstruct the necessary context merely by reading this
  section."

  If you have a relevant JIRA issue, add a link directly to the issue URL here.
  Example: https://issues.folio.org/browse/MODORDERS-70
 -->
This PR is BE-side for populating field "call number" by this structure: "Effective Call Number Prefix" "Effective Call Number" "Effective Call Number Suffix" "Enumeration" "Chronology" "Volume" 
## Approach
<!--
 How does this change fulfill the purpose? It's best to talk
 high-level strategy and avoid code-splaining the commit history.

 The goal is not only to explain what you did, but help other
 developers *work* with your solution in the future.
-->

#### TODOS and Open Questions
<!-- OPTIONAL
- [ ] Use GitHub checklists. When solved, check the box and explain the answer.
-->
The mechanism for fields: Effective Call Number Prefix, Effective Call Number, Effective Call Number Suffix was already created in scope of this ticket: https://issues.folio.org/browse/MODINVSTOR-357

So, the main goal of this story - is to return "enumeration", "chronology" and "volume" in item object, when loans retrieving.
## Learning
<!-- OPTIONAL
  Help out not only your reviewer, but also your fellow developer!
  Sometimes there are key pieces of information that you used to come up
  with your solution. Don't let all that hard work go to waste! A
  pull request is a *perfect opportunity to share the learning that
  you did. Add links to blog posts, patterns, libraries or addons used
  to solve this problem.
-->

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

bohdan-suprun commented on src/test/java/api/support/builders/ItemBuilder.java:
```diff
@@ -422,11 +438,33 @@ public ItemBuilder withYearCaption(List<String> yearCaption) {
         this.temporaryLoanTypeId,
         this.enumeration,
-        this.copyNumbers,
+      this.chronology,
```
Fix alignment

VRohach replied:
Thanks!

bohdan-suprun commented on ramls/loan.json:
```diff
@@ -153,6 +153,18 @@
             }
           }
         },
+        "enumeration": {
```
Please, update API version.

VRohach replied:
@bohdan-suprun do you mean "circulation" version in "ModuleDescriptior-template.json"-file?

bohdan-suprun replied:
I mean circulation RAML and module-descriptor.

VRohach replied:
Ok, thank you for the answer. Done.

bohdan-suprun commented on src/test/java/api/support/builders/ItemBuilder.java:
```diff
@@ -213,7 +220,7 @@ public ItemBuilder withTemporaryLocation(UUID locationId) {
       this.permanentLoanTypeId,
       this.temporaryLoanTypeId,
-      this.enumeration,
+      this.enumeration,this.chronology,
```
```suggestion
      this.enumeration,
      this.chronology,
```

VRohach replied:
I will do it

bohdan-suprun commented on src/main/java/org/folio/circulation/domain/representations/ItemSummaryRepresentation.java:
```diff
@@ -32,6 +32,9 @@ public JsonObject createItemSummary(Item item) {
     write(itemSummary, "barcode", item.getBarcode());
     write(itemSummary, "contributors", item.getContributorNames());
     write(itemSummary, "callNumber", item.getCallNumber());
+    write(itemSummary, "enumeration", item.getEnumeration());
```
The `ItemSummaryRepresentation.java` class is used by various endpoints, e.g.: requests, reports, etc., in a fact they all should be updated as well, or an custom representation should be introduced for loans only (I don't think it make big sense, because we're applying similar changes in https://github.com/folio-org/mod-circulation/pull/429/files to hold-shelf report and requests).

VRohach replied:
Done.

marcjohnson-kint replied:
This topic suggests that other schemas should have changed, and it is marked resolved. Have the changes made to the other schema's been pushed and they don't appear to be in the pull request at the moment?

VRohach commented:
@bohdan-suprun Could you please review and approve this PR?

VRohach commented:
> Given the last merge had also changed the interface versions, these need incrementing again in this pull request.

Ok. I will do it

VRohach commented:
> This change adds these new properties into a number of different scenarios, e.g. checking out an item, creating a loan, fetching an individual loan, fetching multiple loans and anywhere else either the loan schema is used or the `ItemSummaryRepresentation` class is used.
> 
> These properties need adding in to the appropriate tests for these scenarios.
> 
> And there needs to be investigation as to where the `ItemSummaryRepresentation` is used, to check if other schema and scenarios need to be updated.

@marcjohnson-kint I investigated and found that ItemSummaryRepresentation is used in loans, check-in, check-out, requests. So, I added verifying for these new fields in these scenarios. Moreover, as far as I know, @bohdan-suprun  already created tests on it for the requests, so I added only to the loans, check-in, check-out scenarios. 

@marcjohnson-kint Am I right in my investigations? Please, could you review and approve this PR if everything is fine?

marcjohnson-kint commented on src/test/java/api/loans/CheckInByBarcodeTests.java:
```diff
@@ -67,7 +67,11 @@ public void canCloseAnOpenLoanByCheckingInTheItem()
 
     final IndividualResource nod = itemsFixture.basedUponNod(
-      builder -> builder.withTemporaryLocation(homeLocation.getId()));
+      builder -> builder
```
```suggestion
      item -> item
```

marcjohnson-kint commented on src/test/java/api/loans/CheckInByBarcodeTests.java:
```diff
@@ -306,7 +319,11 @@ public void canCheckInAnItemWithoutAnOpenLoan()
 
     final IndividualResource nod = itemsFixture.basedUponNod(
-      builder -> builder.withTemporaryLocation(homeLocation.getId()));
+      builder -> builder
```
```suggestion
      item -> item
```

marcjohnson-kint commented on src/test/java/api/loans/CheckInByBarcodeTests.java:
```diff
@@ -347,7 +373,11 @@ public void canCheckInAnItemTwice()
 
     final IndividualResource nod = itemsFixture.basedUponNod(
-      builder -> builder.withTemporaryLocation(homeLocation.getId()));
+      builder -> builder
```
```suggestion
      item -> item
```

marcjohnson-kint commented on src/test/java/api/loans/CheckOutByBarcodeTests.java:
```diff
@@ -64,7 +64,13 @@ public void canCheckOutUsingItemAndUserBarcode()
-    IndividualResource smallAngryPlanet = itemsFixture.basedUponSmallAngryPlanet();
+
+     IndividualResource smallAngryPlanet = itemsFixture.basedUponSmallAngryPlanet(
+      builder -> builder
```
```suggestion
      item -> item
```

marcjohnson-kint commented on src/test/java/api/loans/CheckOutByBarcodeTests.java:
```diff
@@ -625,7 +640,11 @@ public void canCheckOutUsingUserBarcodeThatContainsSpaces()
 
-    IndividualResource smallAngryPlanet = itemsFixture.basedUponSmallAngryPlanet();
+    IndividualResource smallAngryPlanet = itemsFixture.basedUponSmallAngryPlanet(
+      builder -> builder
```
```suggestion
      item -> item
```

marcjohnson-kint commented on src/test/java/api/loans/CheckOutByBarcodeTests.java:
```diff
@@ -661,7 +689,11 @@ public void canCheckOutUsingProxyUserBarcodeThatContainsSpaces()
 
-    IndividualResource smallAngryPlanet = itemsFixture.basedUponSmallAngryPlanet();
+    IndividualResource smallAngryPlanet = itemsFixture.basedUponSmallAngryPlanet(
+      builder -> builder
```
```suggestion
      item -> item
```

marcjohnson-kint commented on src/test/java/api/loans/LoanAPITests.java:
```diff
@@ -70,7 +70,13 @@ public void canCreateALoan()
 
-    UUID itemId = itemsFixture.basedUponSmallAngryPlanet().getId();
+    IndividualResource smallAngryPlanet = itemsFixture.basedUponSmallAngryPlanet(
+      builder -> builder
```
```suggestion
      item -> item
```

marcjohnson-kint commented on src/test/java/api/loans/OverrideCheckOutByBarcodeTests.java:
```diff
@@ -47,7 +49,12 @@ public void canOverrideCheckoutWhenItemIsNotLoanable()
 
-    IndividualResource smallAngryPlanet = itemsFixture.basedUponSmallAngryPlanet();
+    IndividualResource smallAngryPlanet = itemsFixture.basedUponSmallAngryPlanet(
+      builder -> builder
```
```suggestion
      item -> item
```

VRohach commented:
@marcjohnson-kint Thank you for the review!
